src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1199 @@
+    unordered_map<int32, size_t>
+        initial_arc_map; // the incoming states of these arcs are initial states
+    initial_arc_map.reserve(std::min((int32)1e5, config_.max_active));
+    for (StateIterator<CompactLattice> siter(clat); !siter.Done(); siter.Next()) {
+      auto s = siter.Value();
+      StateId state_append = -1;
+      if (last_get_lattice_frame_ == 0 ||
+          s != 0) { // do not need to copy initial state
+        state_append = s + state_offset;
+        KALDI_ASSERT(state_append == olat->AddState());
+        olat->SetFinal(state_append, clat.Final(s));
+      }
+
+      for (ArcIterator<CompactLattice> aiter(clat, s); !aiter.Done(); aiter.Next()) {
+        const auto &arc = aiter.Value();
+        // construct a copy of the state & arcs
+        if (last_get_lattice_frame_ == 0 ||
+            s != 0) { // do not need to copy initial arc
+          CompactLatticeArc arc_append(arc);
+          arc_append.nextstate += state_offset;
+          olat->AddArc(state_append, arc_append);
+        }
+        if (arc.olabel > config_.max_word_id) {
+          if (s == 0) { // initial_arc
+            initial_arc_map[arc.olabel] = aiter.Position();
+          } else { // final_arc
+            KALDI_ASSERT(clat.Final(arc.nextstate) != CompactLatticeWeight::Zero());
+            final_arc_list_.push_back(
+                pair<int32, size_t>(state_append, aiter.Position()));
+          }
+        }
+      }
+    }
+
+    // step 3.2: connect the states between two chunks
+    if (last_get_lattice_frame_ != 0) {
+      KALDI_ASSERT(final_arc_list_prev_.size());
+      StateId prev_final_state = -1;
+      for (auto &i : final_arc_list_prev_) {
+        MutableArcIterator<CompactLattice> aiter(olat, i.first);
+        aiter.Seek(i.second);
+        auto &arc_append = aiter.Value();
+        auto r = initial_arc_map.find(arc_append.olabel);
+        if (r != initial_arc_map.end()) {
+          ArcIterator<CompactLattice> aiter_chunk(clat, 0); // initial state
+          aiter_chunk.Seek(r->second);
+          const auto &arc_chunk = aiter_chunk.Value();
+          KALDI_ASSERT(arc_chunk.olabel == arc_append.olabel);
+          StateId state_append = arc_chunk.nextstate + state_offset;
+          if (prev_final_state == -1)
+            prev_final_state = arc_append.nextstate;
+          else
+            KALDI_ASSERT(arc_append.nextstate == prev_final_state);
+          CompactLatticeArc arc_append_mod(arc_append);
+          arc_append_mod.nextstate = state_append;
+
+          CompactLatticeWeight weight_offset;
+          weight_offset.SetWeight(
+              LatticeWeight(0, -state_label_forward_prob_[arc_append.olabel]));
+          vector<CompactLatticeWeight> weights = {
+              arc_append_mod.weight, arc_chunk.weight, olat->Final(prev_final_state),
+              weight_offset};
+          BaseFloat v1 = 0, v2 = 0;
+          for (auto &i : weights) v1 += i.Weight().Value1();
```

Comment:
you could have just used Times() on the weights to multiply them, it would have been simpler, if slightly less efficient.